README.md
Queryable JS
------------

Queryable - A tiny, single file database written in Javascript, that emulates a functional subset of mongodb commands. It works in both browser and server. The database itself is stored as a json string. Unlike mongo there are no collections; a separate db is simply returned by each call to open(). Essentially, queryable facilitates structured querying of an array of objects.


## Examples

### Some examples of using queryable.js in node.js:

```js
var queryable = require( 'queryable' );

// There are multiple ways to open a db.
// The simplest is to provide a full path to where you want your database file.
var db = queryable.open( "~/code/mydata.db" );

// ..or you can simply name it, which will start an empty db with this name:
var db = queryable.open( "Database_Name" ); 

// You can load from a json object that is an Array of Objects.
var db = queryable.open( {db_name:"name",data: [{key:val,key2:val2},{key:val},...] } );

// insert any type of key:value pairs you want
db.insert( {key:"keys can be Anything", comment:"fields don't have to match"} );

// Like collections, many independent db can be open at the same time
var db2 = queryable.open( "~/another.db" );

// It handles any value types
db2.insert( {subarray:[1,2,'buckle',{'my':'shoe'}]} );

// find() works like Mongo; RegExp's are fine
var res = db.find( {key:/regex/} );

// SELECT * WHERE (age = 40);
var res = db.find( {age:40} );

// all rows where age is over 40
// supports: $gt, $lt, $gte, $lte, $ne, $exists
var res = db.find( {age: {'$gt':40}} );

// the first 10 rows where age is over 40 and 'name' exists, sorted by name
var res = db.find({age:{$gt:40},name:{'$exists':true}}).sort({name:1}).limit(10);

// find() returns db_result, which has a length property and rows[] array
// as well as chainable methods like: .sort(), .limit(), .skip(), ..
console.log( 'got ' + res.length + ' rows' );


/*
 * a real example - populate from string
 */
// literal data can be a string or an object 
var json_string = '[
  {"name":"Cathy"},
  {"name":"Carol","sex":"f"},
  {"name":"John","sex":"m"},
  {"name":"Cornelius","sex":"m"}]';

var queryable = require('queryable');

var db = queryable.open({db_name:"MyDatabase",data:json_string}); 

// delete a row
db.remove({name:'Cathy'});

// get names that start with 'C'
db.find({name:/^C/}, function(res) {
  console.log( db.db_name + ' contains these names that start with C:' );
  res.rows.forEach(function(x){
    console.log(' ' + x.name);
  });
});

/* outputs:
MyDatabase contains these names that start with C:
 Carol
 Cornelius
*/
```

Callbacks work now too (finally):

```js
  db.distinct('name',{age:{$lte:35}},function(res) {
    res.rows.forEach(function(row){
      console.log( row.name );
    });
  });
```

See the 'examples' folder for more examples of usage.

## Install
To install using npm: 

```
npm install queryable
```
To install in your project, appending to your project's `package.json`: 

```
npm install queryable --save
```

To install using git:

```
~$ git clone https://github.com/gmn/queryable.git
```

## Test

To run unit-tests: 

```
npm test
```


## API

### The queryable.open() function
There are essentially 3 forms of open(): no args, a single string, a single object. The most concise way to specify what you want is to use a `config` object, and specify each param individually. Failing that, merely providing a fullpath string, where you want your database to live, is sufficient in most cases.

#### queryable.open()
* `db_name` = defaults to '*test.db*' in node, '*queryable*' in web browser
* `db_dir`  = default is current working directory: (eg. /home/kooldude)
* `db_path` = concatenation of: (db_dir + db_name) (eg. /home/kooldude/test.db)

#### queryable.open(string)
* First it tries string as a fullpath; if file is found: `db_name`, `db_dir`, and `db_path` are set.
* If a file isn't found, it tries string as a directory where it will create the database using the default `db_name`.
* If it isn't a directory or an existing file, string is assumed to be a fullpath where the file will be saved to. `db_path`, `db_dir`, and `db_name` are set from that.

#### queryable.open(object)
If argument is an object, these configuration variable are looked for in it:
* **db_name**       If none supplied, it either takes one from db_path or defaults to '*queryable*' or '*test.db*' in the browser or in node respectively.
* **db_path**       The fullpath where the database lives. Irrelevant in the browser. The database file is written to db_path everytime save() is called, and read from it in open() if an existing file is discovered at the location.
* **db_dir**        Defaults to current working directory if db_path or db_dir are not supplied.
* **use_gzip**      Can manually specify file is of type gzip. This is useful for files lacking the '.gz' extension. Note: currently gzbz must be present for this to work.
* **data**          If this is set, the database can be populated outright by an argument to data. data can be string or json. It must be in the form of an **Array of Zero or More Objects of Any Kind**.


## Full List of Commands

### Supported Methods
| *object* | command | status | comment/example |
| --- | --- | --- | --- |
| **queryable** | | | the return of `require('queryable')` |
| | `open();` | implemented | returns db_object |
| | `useGzip(boolean);` | implemented | force use of gzip. Useful for gzipped files lacking '.gz' extension; defaults to false |
| **db_object** | | | the return of `queryable.open()` |
| | `save()` | implemented | writes to current system (either file-system or browser) |
| | `insert(vals_obj)` | implemented | inserts vals_obj onto end of table; if vals_obj is an array, will treat contents as array of objects and insert each, one at a time |
| | update(match, newval) | implemented | finds rows that match `match` and set to `newval` |
| | `find(match, callback)` | implemented | returns db_result unless callback is present, then db_result passed to it |
| | `distinct(string,clause)` | implemented | finds all rows, eliminates duplicate values of certain key. `db.distinct('name')` or `db.distinct('name',{age:{$lt:35}})` |
| | `remove()` | implemented | `db.remove( {name:/[A-Z](.*)/} )` |
| | `get_json()` | implemented | returns json string of entire db |
| | `print(fmt)` | implemented | spits out tabular json to stdout |
| | `now()` | implemented | returns Date().toISOString of the present moment; useful for date fields |
| | `toDate(arg)` | implemented | returns new Date(arg) |
| | `count()` | implemented | returns table length |
| **db_result** | | | |
| | `sort()` | implemented | &nbsp; |
| | `limit(integer)` | implemented | limits the result set to `integer` results |
| | `skip(integer)` | implemented | shaves off the first `integer` rows of the result set |
| | `count()` | implemented | returns length of result set |
| | `getArray()` | implemented | &nbsp; |
| | `get_json()` | implemented | &nbsp; |
| | `print(fmt)` | implemented | debug print to stdout |

### Query Operators
| Name | Status | comment/example |
| --- | --- | --- |
| $gt | implemented | `db.find( {a:{$gt:5} );` |
| $gte | implemented | `db.find( {a:{$gte:100} );` |
| $in | *not impl* | selects the documents where a field equals any value in an array |
| $lt | implemented | |
| $lte | implemented | |
| $ne | implemented | Matches all values that are not equal to the value specified in the query `db.find( {a:{$ne:2},b{$ne:3}} ); // select * from table where (a != 2) && (b != 3)` | 
| $nin | *not impl* | |
| | | |
| $or | implemented | `db.find( {$or:[{n:1},{y:{$gte:3}}]} ) //return all results where (n == 1) or (y >= 3)` |
| $and | *not impl* | |
| $not | *not impl* | |
| $nor | *not impl* | |
| | | |
| $exists | implemented | `db.find( {name: {$exists:true} } ) // return all results where the name field exists` |
| $type | *not impl* | |
| | | |
| $mod | no plans | |
| $regex | no plans | |
| $text | no plans | |
| $where | no plans | &nbsp; |

### Update Operators
| Name | Status | comment/example |
| --- | --- | --- |
| $inc | *not impl* | Increments the value of the field by the specified amount. |
| $mul | *not impl* | Multiplies the value of the field by the specified amount. |
| $rename | *not impl* | Renames a field. |
| $setOnInsert | *not impl* | Sets the value of a field upon document creation during an upsert. Has no effect on update operations that modify existing documents. |
| $set | implemented | Sets the value of a field in an existing document. |
| $unset | *not impl* | Removes the specified field from an existing document. |
| $min | *not impl* | Only updates if the existing field value is less than the specified value. |
| $max | *not impl* | Only updates if the existing field value is greater than the specified value. |
| $currentDate | *kindof impl* | Use **db.now()** instead for now |


## In the Works
1. Better documentation (still working on it); Website (not quite yet).
 
2. Feature Roadmap: which features will be implemented next, roughly in order.
  * [X] better documentation for open() 
  * [X] callbacks
  * [] .findOne() 
  * [] better documentation for update() 
  * [] $in, $nin
  * [] $and, $not, $nor
  * [] $inc, $mul 
  * [] $unset 
  * [] $rename 
  * [] update().limit()
  * [] remove().limit()
  * [] $min, $max
  * [] $currentDate    (db.now() already accomplishes this, though could alias it to $currentDate)
  * [] queryable.min.js
  * [] cleanup

3. Sample projects built on queryable demonstrating features and functionality.
  * [Simple Browser Example](http://fatolsun.com/queryable-demo/) (sourceode [here](https://raw.githubusercontent.com/gmn/queryable/master/examples/browser.html))
  * [Arcane Vocabulary Tutor](http://lit-tundra-5131.herokuapp.com/)


## License
(The MIT License)

Copyright 2014 Greg Naughton. All rights reserved.

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


## Contact 
Mail: greg@naughton.org for questions, comments, bugs


changelog.txt
CHANGELOG

==================================================================

V0.3.0

  Major Release: v0.3.0. Not backwards compatible with v0.2.9 or before. 

List of changes:

- changed '_data' member of db_result to 'rows', to make it less obscure.
- tidied the documentation
- added callbacks to:
   - db.save()
   - db.insert()
   - db.update()
   - db.find()
   - db.distinct()
   - db.remove()
- audited and revised error handling in callback bearing functions.
    These now "throw" on any errors resulting from bad input
- db.insert() now returns number of rows inserted, or throws Error for 
    malformed inserts. This way the client application sees malformed inserts immediately, instead of allowing for the chance they miss them, and propagate hard to find bad data in the client application.

==================================================================


examples/exists_test.js

var p = function(s) { console.log(s); };
var json = JSON.stringify;
var queryable = require('../queryable');
var db = queryable.open();
db.insert( [{movie:"Jaws",director:"Stephen Spielberg"},{movie:"Stalker",director:"Andrej Tarkovski",rating:"5/5"},{movie:"Caddy Shack",rating:"4/5"}] );

var res = db.find();
p(json(res));

res = db.find( {director:{$exists:true}} );
p("\n// director exists true\n"+json(res));

res = db.find( {director:{$exists:false}} );
p("\n// director exists false\n"+json(res));

res = db.find( {rating:{$exists:true}} );
p("\n// rating exists true\n"+json(res));

res = db.find( {rating:{$exists:false}} );
p("\n// rating exists false\n"+json(res));


examples/callback_test.js
// utility
var p = function(s) { console.log(s); };

var queryable = require( '../queryable.js' );
var db_name = './callback_test_db.json';
var db = queryable.open( db_name );

// these methods accept callbacks

// - db.insert()
db.insert( [
  {president:"George Washington",took_office:1789},
  {president:"John Adams",took_office:1797},
  {president:"Thomas Jefferson",took_office:1801},
  {president:"James Madison",took_office:1809}
], function(res) {
  p( "inserted " + res + ' rows' );
});

// this will fail
try {
  db.insert( "a string", function(res) {
    p( "inserted " + res + ' rows' );
  });
} catch(e) {
  p( e );
}
// so will this fail
try {
  db.insert( [{row1:1},{row2:1},"not-object"], function(res) {
    p( "inserted " + res + ' rows' );
  });
} catch(e) {
  p( e );
}
// so will this fail
try {
  db.insert( {row1:1},{row2:1}, function(res) {
    p( "inserted " + res + ' rows' );
  });
} catch(e) {
  p( e );
}

// - db.save()
try {
  db.save( function(res) {
    p( "save() returned " + res );
  });
} catch(e) {
  p( e );
}

// - db.find()
db.find( {president:{$exists:true},took_office:{$lte:1801}}, function(res) {
    if ( res.length === 0 ) {
        p( "no rows" );
    } else {
      p( "got " + res.length + ' rows' );
      res.rows.forEach( function( row ) {
          p( row.president + "\t" + row.took_office );
      });
    }
});
  
// - db.update()
db.update( {president:"Bob Dylan"}, {$set:{president:"Bob Dylan", took_office:1965}}, {upsert:true}, function(res) {
    p( res + ' row(s) updated' );
    p( JSON.stringify(db.master) );
});

// - db.distinct()
db.distinct( 'president', null, function(res) {
    p( JSON.stringify(res) );
});

// - db.remove
db.remove( {president:{$exists:false}}, function(res) {
    p( "remove returned: " + res );
    p( JSON.stringify(db.master) );
});



// cleanup
var fs = require('fs');
fs.unlink( db_name );


update_readmes.js
// this script updates the readme attribute of the package.json, from README.md

var fs = require('fs');
var f = fs.readFileSync('package.json',{encoding:'utf8'});
var o = JSON.parse(f);
var package_json_README = o.readme;
var README = fs.readFileSync('README.md',{encoding:'utf8'});

/// expecting a CALLBACK(), and, space-separated string like an actual commandline
function system( callback, commandline )
{
  if ( arguments.length < 2 )
    return;

  var args = '';
  for ( var i = 1; i < arguments.length; i++ ) {
    if ( arguments[i] instanceof Array )
      args += ' ' + arguments[i].join(' ');
    else
      args += ' ' + arguments[i];
  }
  args = args.trim().split(/\s+/);
  
  var res = [];
  var spawn = require('child_process').spawn;
  var child = spawn( args[0], args.slice(1), { env: process.env });
  child.stdout.setEncoding('utf8');
  child.stderr.setEncoding('utf8');

  child.stdout.on('data', function (data) {
      res.push( data );
  });

  child.stderr.on('data', function (data) {
      process.stdout.write( 'stderr: ' + data );
  });

  child.on('close', function (code) {
      //return callback( JSON.stringify(res) );
      return callback(res.join(' ')); 
  });
}
// write readme and diff against it, then remove it
var pkg_readme = 'tmp_' + (Math.random()+'').slice(-4) + '.txt';

var finish_callback = function(s) { 
  process.stdout.write(s); 
  fs.unlinkSync( pkg_readme ); 

  if ( s.trim().length > 0 ) {
    // write newer package.json using README.md 
    process.stdout.write( "writing: \"backup.package.json\"\n" );
    fs.writeFileSync('backup.package.json',f,{encoding:'utf8'});
    o.readme = README;
    process.stdout.write( "overwriting: \"package.json\"\n" );
    fs.writeFileSync( "package.json", JSON.stringify(o,null,'  '), {encoding:'utf8'} );
  } else {
    process.stdout.write( "no change\n" );
  }
};

// write file to diff
fs.writeFileSync( pkg_readme, package_json_README, {encoding:'utf8'} );

// do diff and then finish
system( finish_callback, "diff README.md " + pkg_readme );



examples/populate_with_various_kinds_of_data.js

// a string
var str = '[{"name":"Cathy"},{"name":"Carol","sex":"f"},{"name":"John","sex":"m"}]';

// an object
var obj = [{"name":"Cathy"},{"name":"Carol","sex":"f"},{"name":"John","sex":"m"}];

var queryable = require( '../queryable.js' );

// open, populating (loading) from a string
var db = queryable.open( {"data":str} );

console.log( "populating from String:" );
db.print(1);
db.find().sort( {"name":-1} ).print(3);

// 
console.log( "\npopulating from Object: " );

db = queryable.open( {"data":obj} );
db.print("\t");
db.find({sex:{$exists:true}}).sort( {"sex":-1} ).print();

//--------------- populate entire object from config --------------
obj = {"db_name":"MyCrazyDBName","data":[{"name":"Greg","age":41},{"name":"Ivette","age":35},{"name":"Tucker","age":36}]};
db = queryable.open(obj);
console.log( "Database name is: " + db.db_name + ' and has these people in it:' );
db.find().rows.forEach(function(i){
    console.log(' -> ' + i.name + ' - ' + i.age);
});


examples/unit-min.js

//---------------------------------------------------------------
function UnitTest() {
    this.num_tests = 0;
    this.num_failed = 0;
    this.fs = require('fs');
    var that = this;
}

UnitTest.prototype = {
    stringify: JSON.stringify,
    p: function(s) { console.log(s); },
    po: function(s) { process.stdout.write(s); },
    exe: function(str) { this.p(' '+ str + ' --> ' + eval(str) ); },

    ob_eq: function ( A, B ) {
        return this.stringify(A) === (B + '').trim();
    },

    test: function(str,ans) { 
        ++this.num_tests;
        this.po( "test: \"" + str + '" => '+ans );
        var ret = eval(str);
        if ( this.ob_eq( ret, ans ) ) {
            this.p( " -- Passed" );
        } else {
            this.p( " -- Failed \n"+ans+"\n"+this.stringify(ret) );
            ++this.num_failed;

            var debug = function() {
                eval(str);
            };
            debug();
        }
    },

    // dont evaluate, just compare
    testEq: function(str,ans) {
        ++this.num_tests;
        this.po( "test: \"" + str + '" => '+ans );
        if ( typeof str !== "String" && isNaN(str) )
            str = this.stringify(str);
        if ( str === ans ) {
            this.p( " -- Passed" );
        } else {
            this.p( " -- Failed \n"+ans+"\n"+str );
            ++this.num_failed;
        }
    },

    report: function() {
        this.p( "\n=======================\n * " + this.num_tests + " tests passed successfully" );
        if ( this.num_failed > 0 ) 
            this.p( " ** There were "+this.num_failed+" errors!!!" );
        this.p( "=======================\n" );
    }
};
//---------------------------------------------------------------

debugger;

var db_name = "QUERYABLE-UNITTEST.db";
var U = new UnitTest();
process.on('exit',function() { U.report(); });

if ( U.fs.existsSync( db_name ) ) {
    U.p( '"'+db_name+'" found. Removing...' );
    U.fs.unlink( db_name );
}

var i = 0;
var e = U.fs.existsSync( db_name );
while ( e ) {
    if ( i++ >= 10 ) 
        break;
    e = U.fs.existsSync( db_name );
}
if ( U.fs.existsSync( db_name ) === false ) {
    U.p( '"'+db_name+'" doesn\'t exist. creating...' );
}
else
    U.p( "STILL EXISTS" );

var queryable = require('../queryable.min.js');
var db = queryable.open( db_name );

U.p( "\nPATHS: " );

U.p( "\nINSERT TESTS:" );
U.test( "db.insert({a:1})", 1 );
U.test( "db.insert([{a:2,b:2},{a:3,b:3,c:3}])", 2 );
U.test( "db.insert({a:4,b:4,c:4,d:4})", 1 );
U.test( "db.insert([{b:5,c:5,d:5},{c:6,d:6},{c:7}])", 3 );

U.p( "\nFIND TESTS:" );
U.p( "simple find:" );
U.test( "db.find({a:1})", '{"length":1,"rows":[{"_id":1,"a":1}]}' );
U.p( "simple $gt:" );
U.test( "db.find({b:{'$gt':4}})", '{"length":1,"rows":[{"_id":5,"b":5,"c":5,"d":5}]}' );
U.p( "simple $lt:" );
U.test( "db.find({b:{'$lt':4}})", '{"length":2,"rows":[{"_id":2,"a":2,"b":2},{"_id":3,"a":3,"b":3,"c":3}]}' );
U.p( "simple $gte:" );
U.test( "db.find({b:{'$gte':4}})", '{"length":2,"rows":[{"_id":4,"a":4,"b":4,"c":4,"d":4},{"_id":5,"b":5,"c":5,"d":5}]}' );
U.p( "simple $lte:" );
U.test( "db.find({b:{'$lte':4}})", '{"length":3,"rows":[{"_id":2,"a":2,"b":2},{"_id":3,"a":3,"b":3,"c":3},{"_id":4,"a":4,"b":4,"c":4,"d":4}]}' );
U.p( "reverse sort, numeric:" );
U.test( "db.find({b:{'$gte':4}}).sort({'_id':-1})", '{"length":2,"rows":[{"_id":5,"b":5,"c":5,"d":5},{"_id":4,"a":4,"b":4,"c":4,"d":4}]}' );
U.test( "db.insert({name:'Paul'})", 1 );
U.test( "db.insert({name:'Carol'})", 1 );
U.test( "db.insert({name:'Zach'})", 1 );

U.p( "\nALPHABETICAL SORT:" );
U.test( "db.find({name:/(.*)/}).sort({name:1})", '{"length":3,"rows":[{"_id":9,"name":"Carol"},{"_id":8,"name":"Paul"},{"_id":10,"name":"Zach"}]}' );
U.p( "$exists:true:" );
U.test( "db.find({name:{'$exists':true}}).sort({name:-1})", '{"length":3,"rows":[{"_id":10,"name":"Zach"},{"_id":8,"name":"Paul"},{"_id":9,"name":"Carol"}]}' );
U.p( "\n$exists:false && $exists:false:" );
U.test( "db.find({name:{'$exists':false},'a':{'$exists':false},'b':{'$exists':false}}).sort({_id:-1})", '{"length":2,"rows":[{"_id":7,"c":7},{"_id":6,"c":6,"d":6}]}' );

U.p( "\nAND $exists:true:" );
U.test( "db.find({a:{$exists:true},b:{$exists:true},c:{$exists:true},d:{$exists:true}})", '{"length":1,"rows":[{"_id":4,"a":4,"b":4,"c":4,"d":4}]}' );
U.p( "\nOR:" );
U.test( "db.find({$or:[{d:{$exists:true}},{c:{$exists:true}}]}).sort({_id:-1})", '{"length":5,"rows":[{"_id":7,"c":7},{"_id":6,"c":6,"d":6},{"_id":5,"b":5,"c":5,"d":5},{"_id":4,"a":4,"b":4,"c":4,"d":4},{"_id":3,"a":3,"b":3,"c":3}]}' );

U.p( "\nAND CONDITIONAL:" );
U.test( "db.find({b:{$gt:3},c:{$lte:5},d:{$lt:6},b:{$gte:5}})", '{"length":1,"rows":[{"_id":5,"b":5,"c":5,"d":5}]}' );

U.p("\n\n==========================");
U.p( JSON.stringify(db.find()) );
U.p("==========================\n\n");


U.p( "\nNOT-EQUAL::" );
U.test( "db.find({b:{$ne:4}})", '{"length":9,"rows":[{"_id":1,"a":1},{"_id":2,"a":2,"b":2},{"_id":3,"a":3,"b":3,"c":3},{"_id":5,"b":5,"c":5,"d":5},{"_id":6,"c":6,"d":6},{"_id":7,"c":7},{"_id":8,"name":"Paul"},{"_id":9,"name":"Carol"},{"_id":10,"name":"Zach"}]}' );
U.test( "db.find({b:{$ne:5}})", '{"length":9,"rows":[{"_id":1,"a":1},{"_id":2,"a":2,"b":2},{"_id":3,"a":3,"b":3,"c":3},{"_id":4,"a":4,"b":4,"c":4,"d":4},{"_id":6,"c":6,"d":6},{"_id":7,"c":7},{"_id":8,"name":"Paul"},{"_id":9,"name":"Carol"},{"_id":10,"name":"Zach"}]}' );
U.test( "db.find({a:{$ne:3},b:{$ne:2},c:{$ne:6}})", '{"length":7,"rows":[{"_id":1,"a":1},{"_id":4,"a":4,"b":4,"c":4,"d":4},{"_id":5,"b":5,"c":5,"d":5},{"_id":7,"c":7},{"_id":8,"name":"Paul"},{"_id":9,"name":"Carol"},{"_id":10,"name":"Zach"}]}' );



U.p( "\nOR CONDITIONAL:" );
U.test( "db.find( {a:{$exists:true},b:{$lt:3},a:{$gte:1} } )", '{"length":1,"rows":[{"_id":2,"a":2,"b":2}]}' );
U.test( "db.find( {b:{$gt:4},c:{$gte:5},d:{$lte:6}})", '{"length":1,"rows":[{"_id":5,"b":5,"c":5,"d":5}]}' );
U.test( "db.find( {a:3,b:{$gt:2},c:{$lt:7}} )", '{"length":1,"rows":[{"_id":3,"a":3,"b":3,"c":3}]}' );

U.p( "\nAND/OR CONDITIONAL COMBINED:" );
U.test( "db.find( { c:{$lt:6}, $or: [{a:{$gt:5}},{b:{$lte:4}}] } )", '{"length":2,"rows":[{"_id":3,"a":3,"b":3,"c":3},{"_id":4,"a":4,"b":4,"c":4,"d":4}]}' );

U.p( "\nREMOVE: ");
U.test( "db.remove( {name:/[A-Z](.*)/} )", 3 );
U.test( "db.find().count()", 7 );
U.test( "db.remove( {a:{$gt:1},b:{$gt:2},c:{$exists:true},d:{$exists:false}} )", 1 );
U.test( "db.count()", 6 ); // count can be called from db object or result object
U.test( "db.remove( {c:{$exists:true}} )", 4 );

U.p( "\nUPDATE: ");
U.test( "db.update( {b:{$exists:true}}, {$set:{awesome:true}}, {multi:true} )", 1 );
U.test( "db.update( {}, {$set:{everything:1}}, {multi:true} )", 2 );
U.p(db.get_json());
U.test( "db.update( {not:'here'}, {$set:{put:'anyway'}}, {upsert:true} )", 1 );
U.p(db.get_json());

U.p( "\nLIMIT: ");
U.test( "db.remove()", 3 );

var names = ["Dan","Fred","Amy","Ivette","Justin","Phoebe"];
for ( var i = 0; i < 100; i++ ) {
    db.insert( {person_id:1000+i,name:names[Math.floor(Math.random()*names.length)]} );
}
var ph = 0;
names.forEach(function(n){
    var r = db.find({name:n}).count();
    U.p( ' --> adding: '+ n + ' ' + r );
    if ( n === "Phoebe" ) ph = r;
});
for ( var lim = 0; lim <= ph+1; lim++ ) {
    if ( lim === ph+1 )
        U.test( "db.find( {name:'Phoebe'} ).limit("+lim+").count()", lim-1 );
    else
        U.test( "db.find( {name:'Phoebe'} ).limit("+lim+").count()", lim );
}

/*
    SKIP
*/

U.p( "\nSKIP: ");
db.remove(); 
var A = [];
for ( A = [], i = 0; i < 100; i++ ) {
    A.push({'i':i});
}
db.insert(A);
U.test("A=[];db.find().skip(40).limit(20).rows.forEach(function(o){A.push(o['i']);});A;", '[40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59]' );

/*
    DISTINCT
*/
U.p( "\nDISTINCT: " );
db.remove();
U.test( "db.find()" , '{"length":0,"rows":[]}' );
db = queryable.open();
db.insert( {tableName:'whatever'} );
A = [];
for ( A = [], i = 0; i < 4; i++ ) {
    A.push( {'a': i%2} );
}
db.insert(A);
U.test("db.find();",'{"length":5,"rows":[{"_id":1,"tableName":"whatever"},{"_id":2,"a":0},{"_id":3,"a":1},{"_id":4,"a":0},{"_id":5,"a":1}]}');
U.test("db.distinct('a');",'{"length":2,"rows":[{"_id":2,"a":0},{"_id":3,"a":1}]}');
db.insert([{'a':2},{'a':3}]);
U.test("db.distinct('a',{'a':{$lt:1}});",'{"length":1,"rows":[{"_id":2,"a":0}]}');
U.test("db.distinct('a',{$or:[{'tableName':{$exists:true}},{'a':{$lt:1}}]});",'{"length":1,"rows":[{"_id":2,"a":0}]}');
U.test("db.distinct('a',{$or:[{'tableName':{$exists:true}},{'a':{$gte:2}}]});",'{"length":2,"rows":[{"_id":6,"a":2},{"_id":7,"a":3}]}');
db.insert([{'a':2},{'a':3}]);
U.test("db.distinct('a',{$or:[{'tableName':{$exists:true}},{'a':{$gte:2}}]});",'{"length":2,"rows":[{"_id":6,"a":2},{"_id":7,"a":3}]}');
var db4 = queryable.open( {data:[{n:'jared'},{n:'martha'},{n:'jim'},{n:'nolan'},{n:'jim'},{n:'jared'}]});
U.test("db4.distinct('n',{n:/^j/});", '{"length":2,"rows":[{"_id":1,"n":"jared"},{"_id":3,"n":"jim"}]}' );

/*
CALLBACKS
*/
U.p("\nCALLBACKS:");
// FIND
db.find(null, function(res) {
  U.testEq( res, '{"length":9,"rows":[{"_id":1,"tableName":"whatever"},{"_id":2,"a":0},{"_id":3,"a":1},{"_id":4,"a":0},{"_id":5,"a":1},{"_id":6,"a":2},{"_id":7,"a":3},{"_id":8,"a":2},{"_id":9,"a":3}]}' );
});

db.remove();
db.find(null, function(res) {
  U.testEq( res, '{"length":0,"rows":[]}' );
});

// INSERT
db = queryable.open();
for ( var i = 0; i < 4; i++ ) {
  db.insert( {k:i}, function(res) {
      U.testEq(res, 1);
  } );
}

// UPDATE
db.update( {k:0}, {$set:{k:4}}, null, function(res) {
  U.testEq(res, 1);
});

// DISTINCT
db.distinct( 'k', null, function(res) {
  U.testEq(res, '{"length":4,"rows":[{"_id":1,"k":4},{"_id":2,"k":1},{"_id":3,"k":2},{"_id":4,"k":3}]}' );
});
db.distinct( 'k', {k:{$gt:2}}, function(res) {
  U.testEq(res, '{"length":2,"rows":[{"_id":1,"k":4},{"_id":4,"k":3}]}' );
});

// REMOVE
db.remove( {k:{$gt:0}}, function(res) {
  U.testEq( res, 4 );
});
